### PR TITLE
Collections

### PR DESCRIPTION
**Pull request by UladBohdan**
- ArticleCollection model
- Collections API
- separate routers for Collections, Brands and Articles
- proper testing for Articles API and Collections API
- an update for Postman configuration
- nodejs version updates (Travis and `package.json`)
- active=false instead of Object removing
- an optional `collectionSlug` on article Create to associate an article with a collection (and vice versa).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
     "express": "^4.15.3",
     "express-session": "^1.15.4",
     "lodash": "^4.17.4",
-    "mongoose": "^4.11.7",
+    "mongoose": "^4.13.8",
```
Comment (Drapegnik) [inline package.json]: please update [`connect-mongo`](https://github.com/jdesboeufs/connect-mongo/blob/master/CHANGELOG.md#200--2017-10-09) too

File: src/api/article/brand/tests.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+describe('Brands API', () => {
+  const brands = ['Wir', 'Kurilka', 'Minsk'];
+
+  before(async () => {
+    // Ensuring DB is free of any ArticleBrands.
+    await ArticleBrand.remove();
+    // Populating DB with Brands.
+    const promises = [];
+    brands.forEach(brand => promises.push(new ArticleBrand({ name: brand }).save()));
```
Comment (Drapegnik) [inline src/api/article/brand/tests.js]: please use `.map` instead

File: src/api/article/brand/tests.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    await Promise.all(promises);
+  });
+
+  after(async () => {
+    const promises = [];
+    brands.forEach(brand => promises.push(ArticleBrand.remove({ name: brand })));
```
Comment (Drapegnik) [inline src/api/article/brand/tests.js]: same as above

File: src/api/article/collection/controller.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    .then(sendJson(res))
+    .catch(next);
+
+export const getOne = ({ params: { slug } }, res, next) =>
+  ArticleCollection.findOne({ slug, active: true })
+    .then(checkIsFound)
+    .then(sendJson(res))
+    .catch(next);
+
+export const create = async ({ body }, res, next) => {
```
Comment (Drapegnik) [inline src/api/article/collection/controller.js]: can you try to rewrite it with promises chain?

File: src/db/init.js [hunk 5]
```diff
@@ -59,6 +61,27 @@ const initArticles = articleBrandsDict =>
...
+  });
+  return articlesDict;
+};
+
+const initArticleCollections = articlesDict =>
+  Promise.all(
+    articleCollectionsData.map(async articleCollectionData => {
+      const articleIds = articleCollectionData.articleSlugs.map(slug => articlesDict[slug]);
+      return new ArticleCollection(
+        assign(omit(articleCollectionData, ['articleSlugs']), { articles: articleIds })
```
Comment (Drapegnik) [inline src/db/init.js]: the same as:

File: src/db/init.js [hunk 6]
```diff
@@ -59,6 +61,27 @@ const initArticles = articleBrandsDict =>
...
+    // eslint-disable-next-line no-underscore-dangle
+    articlesDict[item.slug] = item._id;
+  });
+  return articlesDict;
+};
+
+const initArticleCollections = articlesDict =>
+  Promise.all(
+    articleCollectionsData.map(async articleCollectionData => {
+      const articleIds = articleCollectionData.articleSlugs.map(slug => articlesDict[slug]);
```
Comment (Drapegnik) [inline src/db/init.js]: `Object.values(articlesDict)`

File: src/api/article/collection/controller.js (hunk 4 above)
Comment (UladBohdan) [reply src/api/article/collection/controller.js]: Will try that later.
The rest of your comments are resolved.

**Comment by UladBohdan** [comment]
## Coverage 79% -> 90%

**Comment by UladBohdan** [comment]
Seems to be enough changes for one PR.
@Drapegnik review-approve it please.

File: src/api/article/controller.js [hunk 7]
```diff
@@ -53,14 +58,25 @@ export const create = async ({ body }, res, next) => {
     const articleBrandQuery = ArticleBrand.findOne({ name: body.brand });
     const articleBrand = (await articleBrandQuery.exec()) || new ArticleBrand({ name: body.brand });
     await articleBrand.save();
-    const articleBody = body;
-    articleBody.brand = articleBrand._id; // eslint-disable-line no-underscore-dangle
+
+    const articleCollection = await ArticleCollection.findOne({ slug: body.collectionSlug });
+
+    const articleBody = omit(body, ['collectionSlug']);
```
Comment (Drapegnik) [inline src/api/article/controller.js]: is it added to model `collectionSlug` if not omit it here?

Comment (Drapegnik) [reply src/api/article/controller.js]: if not, that code looks more readable:

Comment (UladBohdan) [reply src/api/article/controller.js]: 1. `omit` here is redundant although we may want to keep it to emphasize `collectionSlug` is not a part of a model.

2. Your approach raises a `property _id of undefined` error.
The code below works:
const articleBody = {
      ...body,
      brand: articleBrand._id,
      collectionId: articleCollection ? articleCollection._id : null,
    };
Although `null` is not the same as nothing: `getAll` or `getOne` than return a `collectionId` property with `null` value while for other objects `collectionId` is not returned at all.

Comment (Drapegnik) [reply src/api/article/controller.js]: better: ```jscollectionId: articleCollection && articleCollection._id```

**Pull request merged**